src/include/migraphx/op/slice.hpp

```diff
@@ -90,80 +190,141 @@ struct slice
+     * \param ax_vec axes to slice on
+     */
+    template <class IndView, class Axes>
+    auto compute_offset(const shape& s, const IndView& input_starts, const Axes& ax_vec) const
     {
-        check_shapes{inputs, *this, true}.has(1);
-        auto input_shape = inputs[0];
-        auto t           = input_shape.type();
-
-        // TODO:  When support for dynamic shapes is added to normalize_attributes,
-        //  remove this restriction.
-        if(input_shape.dynamic() and std::any_of(axes.begin(), axes.end(), [&](auto axis) {
-               return not input_shape.dyn_dims()[axis].is_fixed();
-           }))
+        auto ret = 0;
+        for(std::size_t i = 0; i < ax_vec.size(); ++i)
         {
-            MIGRAPHX_THROW("SLICE: slicing is not allowed on non-fixed dynamic input axis ");
+            auto axis = ax_vec[i];
+            ret += input_starts[i] * s.strides().at(axis);
         }
+        return ret * s.type_size();
+    }
 
-        // For a static shape, old_lens will be adjusted to a new size
-        // for those axes that are sliced.
-        // For dynamic shape, the adjusted old_lens become the new max values,
-        // while updating the old mins and optimals if possible.
-        std::vector<std::size_t> new_mins;
-        std::vector<std::size_t> old_lens;
-        std::vector<std::size_t> old_strides;
-        // Doesn't handle optimals
-        if(input_shape.dynamic())
-        {
-            old_lens = input_shape.max_lens();
-            new_mins = input_shape.min_lens();
-        }
-        else
-        {
-            old_lens = input_shape.lens();
-            // For static shape (including during eval step after a dynamic input) the strides are
-            // indexed into the pre-slice array, so they are larger than the apparent size of the
-            // resulting shape.
-            old_strides = input_shape.strides();
-        }
+    /**
+     * Used on the input_starts and input_ends to make them positive and within bounds.
+     */
+    template <class Inds, class Axes>
+    Inds normalize_indices(shape input_shape, Inds indices, Axes in_axes) const
+    {
+        std::transform(in_axes.begin(),
+                       in_axes.end(),
+                       indices.begin(),
+                       indices.begin(),
+                       [&](auto axis, auto index) {
+                           auto dim = input_shape.lens().at(axis);
+                           if(index < 0)
+                           {
+                               index += dim;
+                           }
+                           if(index < 0)
+                           {
+                               index = 0;
+                           }
+                           else if(index > dim)
+                           {
+                               index = dim;
+                           }
+                           return index;
+                       });
+        return indices;
+    }
```

Comment:
A function like this should go into normalize_attributes.hpp and it should reuse the`tune_attribute` function we have. I would envision a function like this:

```cpp
auto normalize_indices(const std::vector<int64_t> &indices,
                       const std::vector<int64_t> &axes,
                       const shape &input_shape,
                       const std::vector<op::normalize_attribute> &attributes);
``` 

This way we can use the same attributes that we use for the static case.

Reply:
That's possible but very messy to reuse tune_attribute. It  would also have to be a template because the types of `indices` and `in_axes` are not necessarily the same.

Reply:
I don't think we want to do this with how operation.hpp and normalize_attribute.hpp is set up. @pfultz2 What do you think?

Reply:
> That's possible but very messy to reuse tune_attribute.

It shouldn't be messy. I think an implementation like this would work:

```cpp
std::vector<int64_t>
normalize_indices(const std::vector<int64_t> &indices,
                  const std::vector<int64_t> &axes, const shape &input_shape,
                  const std::vector<op::normalize_attribute> &attributes)
{
  return tune_attribute(indices, axes, attributes, input_shape,
                        [] { return ""; });
}
```

In the future we could consider changing `tune_attribute` to take `std::vector<op::normalize_attribute>` instead of `value`, but that doesn't need to be addressed in this API. 

To improve the error message we could add a another `prefix` parameter:

```cpp
std::vector<int64_t>
normalize_indices(const std::vector<int64_t> &indices,
                  const std::vector<int64_t> &axes, const shape &input_shape,
                  const std::vector<op::normalize_attribute> &attributes,
                  const std:string& prefix = "")
{
  return tune_attribute(indices, axes, attributes, input_shape,
                        [&] { return prefix; });
}
```


> It would also have to be a template because the types of indices and in_axes are not necessarily the same.

No need to template it, we can just convert them to `std::vector<int64_t>` to call the function.

Reply:
I tried it out, having issues with converting the `tensor_view` into a `std::vector<int64_t>`:
```
    /**
     * Three input version of the normalize_inputs.
     * This one also checks that the input_axes are valid.
     */
    template <class Inds>
    std::unordered_map<std::string, Inds>
    normalize_inputs(shape input_shape, Inds input_starts, Inds input_ends, Inds input_axes) const
    {
        auto attrs = this->attributes().at("normalize_axes");
        return {{"input_starts", normalize_input(input_starts.to_vector(),
                    this->axes, input_shape, attrs.at("starts"), "Slice variable input_starts")},
                {"input_ends", normalize_input(input_ends.to_vector(),
                    this->axes, input_shape, attrs.at("ends"), "Slice variable input ends")},
                {"input_axes", normalize_input(input_axes.to_vector(),
                    {}, input_shape, attrs.at("axes"), "Slice variable input_axes")}};
    }
```

Reply:
Think I can get around it by using `tensor_view.template to_vector<int64_t>`. Similar thing needs to be done for the converting the attributes value.